src/verifier/mod.rs

```diff
@@ -1,120 +1,83 @@
 #[derive(Debug, PartialEq, Eq)]
-pub enum GuessChar {
-    Hit(char),
-    Miss(char),
-    Move(char),
+pub enum GuessResult {
+    Guessed,
+    Missed { hint: String },
 }
 
 #[derive(Debug)]
-pub struct Verifier {
-    input: Vec<GuessChar>,
-    secret: Vec<char>,
+pub struct Verifier<'a> {
+    secret: &'a str,
 }
 
-impl Verifier {
-    pub fn new(secret: &str) -> Self {
-        Verifier {
-            input: Vec::new(),
-            secret: secret.chars().collect(),
-        }
+impl<'a> Verifier<'a> {
+    pub fn new(secret: &'a str) -> Self {
+        Verifier { secret }
     }
 
-    pub fn hint(&self) -> String {
-        let mut output = String::new();
-        for item in self.input.iter() {
-            match item {
-                GuessChar::Hit(_) => output.push('+'),
-                GuessChar::Miss(_) => output.push('-'),
-                GuessChar::Move(_) => output.push('?'),
-            }
+    pub fn verify(&mut self, input: &str) -> GuessResult {
+        if input == self.secret {
+            return GuessResult::Guessed;
         }
-        output
-    }
-
-    fn parse_input(input: &str) -> Vec<GuessChar> {
-        input.trim().chars().map(GuessChar::Miss).collect()
-    }
 
-    pub fn verify(&mut self, input: &str) -> bool {
-        self.input = Self::parse_input(input);
+        let secret_chars = self.secret.chars().collect::<Vec<_>>();
```

Comment:
Considering that this `Vec` allocation is gonna happen every time you run `verify` I'd say it make sense for the `Verifier` to take ownership of the `secret` (store String instead of &'a secret)

Reply:
It is already owned as I eliminated references and lifetimes in 167ac962b9ff8aae9209feacb6429b6b2b019852